File: LeetCode/Graph/TopSort/py/Leetcode-外星文字典.py
class Solution:
    def alienOrder(self, words: List[str]) -> str:
        mp,n,flag = defaultdict(list),len(words), True
        din = defaultdict(int)
        for word in words:
            for i in range(len(word)):
                c = word[i]
                if c not in mp:
                    mp[c] = []


        def build(l,r):
            nonlocal flag
            n = min(len(l),len(r))
            i =  0
            while i < n:
                a,b = l[i],r[i]
                if a != b:
                    mp[a].append(b)
                    din[b] += 1
                    break
                i += 1
            if i == n and len(l) > len(r):
                flag = False
    
        for i in range(1,n):
            build(words[i - 1],words[i])
            if not flag: break


        if not flag: return ''

        res = []        
        def topsort():
            nonlocal res
            q,cnt = deque(),0
            for k,v in mp.items():
                if k not in din:
                    q.append(k)

            while q:
                t = q.popleft()
                res.append(t)
                cnt += 1
                for ne in mp[t]:
                    din[ne] -= 1
                    if din[ne] == 0:
                        q.append(ne)

            if cnt != len(mp): return True
            return False

        if topsort(): return ''

        return ''.join(res) 

File: LeetCode/Graph/TopSort/js/Leetcode-外星文字典.js
var alienOrder = function (words) {
    let [mp, din, n] = [new Map(), new Map(), words.length];
    let flag = true;
    for (const word of words) {
        for (let i = 0; i < word.length; i++) {
            let c = word[i];
            if (!mp.has(c)) mp.set(c, []);
        }
    }
    let build = function (l, r) {
        const n = Math.min(l.length, r.length);
        let i = 0;
        while (i < n) {
            const [a, b] = [l[i], r[i]];
            if (a != b) {
                mp.get(a).push(b);
                din.set(b, (din.get(b) || 0) + 1);
                break;
            }
            i++;
        }
        if (i == n && l.length > r.length) {
            flag = false;
        }
    };

    for (let i = 1; i < n && flag; i++) {
        build(words[i - 1], words[i]);
    }

    // console.log(din);

    if (!flag) return "";
    let res = [];
    let topsort = function () {
        let q = [];
        for (let k of mp.keys()) {
            if (!din.has(k)) q.push(k);
        }

        let cnt = 0;

        while (q.length) {
            let t = q.shift();
            cnt++;
            // console.log("-----",t);
            res.push(t);
            for (let ne of mp.get(t)) {
                din.set(ne, din.get(ne) - 1);
                // console.log(ne,din.get(ne));
                if (din.get(ne) === 0) {
                    q.push(ne);
                }
            }
        }
        if (cnt != mp.size) return false;
        return true;
    }

    if (!topsort()) return ""
    return res.join('');
};